staging/vhost-device-sound/src/audio_backends/alsa.rs

```diff
@@ -716,4 +718,528 @@ impl AudioBackend for AlsaBackend {
+    }
+
+    #[test]
+    fn test_alsa_invalid_state_transitions() {
+        crate::init_logger();
+        setup_alsa_conf();
+
+        let streams = Arc::new(RwLock::new(vec![
+            Stream::default(),
+            Stream {
+                id: 1,
+                direction: Direction::Input,
+                ..Stream::default()
+            },
+        ]));
+        let request = VirtioSndPcmSetParams {
+            hdr: VirtioSoundPcmHeader {
+                stream_id: 3.into(),
+                hdr: VirtioSoundHeader { code: 0.into() },
+            },
+            format: VIRTIO_SND_PCM_FMT_S16,
+            rate: VIRTIO_SND_PCM_RATE_44100,
+            channels: 2,
+            features: 0.into(),
+            buffer_bytes: 8192.into(),
+            period_bytes: 4096.into(),
+            padding: 0,
+        };
+        {
+            let backend = AlsaBackend::new(streams.clone());
+
+            // Invalid, but we allow it.
+            backend.stop(0).unwrap();
+            // Invalid, but we don't allow it.
+            backend.release(0).unwrap_err();
+            backend.start(0).unwrap_err();
+            backend.release(0).unwrap_err();
+        }
+        {
+            let backend = AlsaBackend::new(streams.clone());
+
+            // set_parameters -> set_parameters | VALID
+            backend.set_parameters(0, request).unwrap();
+
+            // set_parameters -> prepare | VALID
+            backend.prepare(0).unwrap();
+
+            // Invalid, but we allow it.
+            // prepare -> stop | INVALID
+            backend.stop(0).unwrap();
+            // prepare -> release | VALID
+            backend.release(0).unwrap();
+
+            // release -> start | INVALID
+            backend.start(0).unwrap_err();
+            // release -> stop | VALID
+            backend.stop(0).unwrap();
+            // release -> prepare | VALID
+            backend.prepare(0).unwrap();
+
+            // prepare -> start | VALID
+            backend.start(0).unwrap();
+
+            // start -> start | INVALID
+            backend.start(0).unwrap_err();
+            // start -> set_parameters | INVALID
+            backend.set_parameters(0, request).unwrap_err();
+            // start -> prepare | INVALID
+            backend.prepare(0).unwrap_err();
+            // start -> release | INVALID
+            backend.release(0).unwrap_err();
+            // start -> stop | VALID
+            backend.stop(0).unwrap();
+            // stop -> start | VALID
+            backend.start(0).unwrap();
+            // start -> stop | VALID
+            backend.stop(0).unwrap();
+            // stop -> prepare | INVALID
+            backend.prepare(0).unwrap_err();
+            // stop -> set_parameters | INVALID
+            backend.set_parameters(0, request).unwrap_err();
+            // stop -> release | VALID
+            backend.release(0).unwrap();
+        }
+
+        // Redundant checks? Oh well.
```

Comment:
This is a minor comment for the future. I think at some point we could move the PCM state machine to the device and implement it once for all audio backends. In that moment, we can move these tests to device.rs

Reply:
Yeah certainly. These are just to cover the backend's code paths otherwise they were shown as not executed in the tests.